спринт 12 на проверку

PavelRomashin opened the pull request:
спринт 12 на проверку

PavelRomashin committed:
спринт 12 на проверку

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/App.kt:
```diff
@@ -0,0 +1,29 @@
+import android.app.Application
+import androidx.appcompat.app.AppCompatDelegate
+
+const val SETTING_PREFERENCES = "Setting_preferences"
+const val DARK_THEME_KEY = "Dark_theme_key"
```
🍏 Константы лучше объявлять в companion object

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -22,27 +24,37 @@ import retrofit2.Retrofit
 import retrofit2.converter.gson.GsonConverterFactory
 
 
+const val HISTORY_PREFERENCES = "HISTORY_PREFERENCES"
```
🍏 Также можешь в текущем классе добавить в companion object

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -55,6 +67,15 @@ class SearchActivity : AppCompatActivity() {
         errorImage = findViewById(R.id.errorImage)
         errorText = findViewById(R.id.errorText)
         updateButton = findViewById(R.id.buttonUpdate)
+        recyclerViewHistory = findViewById(R.id.search_history_recycler_view)
+        recyclerView = findViewById(R.id.recyclerView)
+        recyclerView.adapter = trackAdapter
+        sharedPrefsHistory = getSharedPreferences(HISTORY_PREFERENCES, MODE_PRIVATE)
+        searchHistory = SearchHistory(sharedPrefsHistory)
+        clearSearchButton = findViewById(R.id.clear_search_history)
+        layoutForHistory = findViewById(R.id.search_history)
+        errorLayout = findViewById(R.id.errorLayout)
```
🍏 Чтобы не загромождать метод onCreate, можно настройку view-компонентов вынести в отдельные функции

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -90,6 +125,14 @@ class SearchActivity : AppCompatActivity() {
             override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
 
                 clearButton.visibility = clearButtonVisibility(s)
+                layoutForHistory.visibility =
+                    if (inputEditText.hasFocus() && s?.isEmpty() == true) View.VISIBLE else View.GONE
+                clearSearchButton.visibility =
+                    if (inputEditText.hasFocus() && s?.isEmpty() == true) View.VISIBLE else View.GONE
```
⚠ Здесь лучше учесть, что список просмотренных треков может быть пустым. Данный кейс происходит, если ввести текст в поле и очистить, не добавив в историю треки

![Снимок экрана 2023-03-12 230247](https://user-images.githubusercontent.com/72722086/224563728-744092f4-d1a9-4ec6-a675-8df47bea2702.png)

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -99,6 +142,33 @@ class SearchActivity : AppCompatActivity() {
+
+
+        trackAdapter.itemClickListener = { track ->
+            if (trackLibraryHistory.contains(track)) {
+                trackLibraryHistory.remove(track)
+                trackLibraryHistory.add(0, track)
+            } else {
+                trackLibraryHistory.add(0, track)
+            }
+            if (trackLibraryHistory.size == 10) {
+                trackLibraryHistory.removeAt(9)
+            }
```
🍏 Вызов trackLibraryHistory.add(0, track) можно здесь делать один раз после всех проверок. В любом случае трек должен быть добавлен в начало списка

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -99,6 +142,33 @@ class SearchActivity : AppCompatActivity() {
+        clearSearchButton.setOnClickListener {
+            searchHistory.clearHistory()
+            trackLibraryHistory.clear()
+            layoutForHistory.visibility = View.GONE
+            clearSearchButton.visibility = View.GONE
```
🍏 Эти view добавлены в общий контейнер и в коде неоднократно приходится регулировать видимость этих компонентов в одинаковых условиях, но по отдельности. Можно присвоить id для этого контейнера и объявить здесь переменную, чтобы скрывать или показывать контейнер целиком

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,26 @@
+
+
+    fun clearHistory() {
+        sharedPrefs.edit().remove(TRACK).apply()
```
🍏 Можно использовать extension для SharedPreferences и удаление/запись объявлять так:
        sharedPrefs.edit { remove(TRACK) }
или так
     sharedPrefs.edit { putString(TRACK, json) }
Вызов apply скрыт в реализации функции

PavelRomashin committed:
Спринт 12. Исправление замечаний
1) В классах App и SearchActivity константы вынесены в companion object.
2) Объявление переменных в классе SearchActivity вынесено в функцию announce.
3) Исправлена ошибка, при которой при очистке "на крестик" оставались кнопка "очистить историю" и текст "вы искали"
4) удалено дублирование добавление трека в SearchActivity.
5) Вызов функции чтения/записи исправлен в соответствии с рекомендацией.

ilshat-abdulin commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -87,6 +76,8 @@ class SearchActivity : AppCompatActivity() {
             errorText.isVisible = false
             updateButton.isVisible = false
+            layoutForHistory.visibility = View.GONE
+            clearSearchButton.visibility = View.GONE
```
⚠ Касаемо прошлого замечания на счет истории поиска. Я специально оставил комментарий в коде onTextChanged, т.к. имел ввиду, что нужно там дополнить проверку для отображения/скрытия layoutForHistory и clearSearchButton. Дополнительно нужно проверять searchHistory на пустоту, по аналогии в setOnFocusChangeListener. Т.к. при срабатывании clearButton.setOnClickListener для inputEditText устанавливается пустая строка и срабатывает onTextChanged в TextWatcher

А это нужно убрать
layoutForHistory.visibility = View.GONE
clearSearchButton.visibility = View.GONE
т.к. после срабатывания setOnClickListener, проявляется некорректное поведение после добавления треков в историю. История не отображается (хотя треки добавлены и поле ввода в фокусе). 
Отображается, только если ввести что-нибудь в поле и вручную очистить без clearButton

PavelRomashin committed:
Спринт 12. Исправление замечаний 2
1) Внес проверку наличия треков в истории в Text Watcher и удалил строки с установкой видимости в clearButton.setOnClickListener. Сейчас всё вроде хорошо работает.
